# Lab 04 - pandas II

Name: Kartikeya Sharma  
Class: CSCI 349 - Intro to Data Mining  
Semester: 2021SP  
Instructor: Brian King  

#### To start, create your first cell. Then, place the usual import statements at the top of your notebook file.

In [201]:
import sys
import numpy as np
import pandas as pd
        
print("Python version: " + str(sys.version))
print()
print("Numpy version: " + str(np.__version__))
print("Pandas version: " + str(pd.__version__))

Python version: 3.8.8 (default, Feb 24 2021, 13:46:16) 
[Clang 10.0.0 ]

Numpy version: 1.19.2
Pandas version: 1.2.3


#### In the 2nd cell, let's create a bigger version of the quiz scores dataset with the following features:
#### * We have two students, Bob and Jane
#### * We have four weeks of data. Notice we can have duplicate names in our index. (However, columns mustbe unique! You can't have duplicate variable names!)
#### * There are missing quizzes, indicated with a zero in those locations.

#### Copy the following into your 2nd cell:

In [202]:
days = ["Mon","Tue","Wed","Thurs","Fri","Student"]
scores = pd.DataFrame([pd.Series([9.5, 8.75, 8, 10, 7.75, "Bob"], index=days, name="week_1"),
pd.Series([9, 8, 10, 8.75, 7.25, "Jane"], index=days, name="week_1"),
pd.Series([8.5, 0, 9.75, 9, 6, "Bob"], index=days, name="week_2"),
pd.Series([7, 8.25, 9.25, 0, 8, "Jane"], index=days, name="week_2"),
pd.Series([7, 8.5, 9.25, 0, 0, "Bob"], index=days, name="week_3"),
pd.Series([8.5, 8.25, 9, 8, 7.5, "Jane"], index=days, name="week_3"),
pd.Series([9, 7, 8.5, 8, 8, "Bob"], index=days, name="week_4"),
pd.Series([9.25, 9.25, 8.5, 7.75, 7.5, "Jane"], index=days,

name="week_4"),
]) 

scores

Mon   Tue    Wed  Thurs   Fri Student
week_1  9.50  8.75   8.00  10.00  7.75     Bob
week_1  9.00  8.00  10.00   8.75  7.25    Jane
week_2  8.50  0.00   9.75   9.00  6.00     Bob
week_2  7.00  8.25   9.25   0.00  8.00    Jane
week_3  7.00  8.50   9.25   0.00  0.00     Bob
week_3  8.50  8.25   9.00   8.00  7.50    Jane
week_4  9.00  7.00   8.50   8.00  8.00     Bob
week_4  9.25  9.25   8.50   7.75  7.50    Jane

#### 1. [P] Report the shape, the number of dimensions, the size of the data, and the data types of each column. Print each separately. Then create an additional code cell that shows the results of ```scores.info()```.

In [203]:
print("shape: " + str(scores.shape))
print("number of dimensions: " + str(scores.ndim))
print("size of data: " + str(scores.size))
print("types of each column: \n\n" + str(scores.dtypes))

shape: (8, 6)
number of dimensions: 2
size of data: 48
types of each column: 

Mon        float64
Tue        float64
Wed        float64
Thurs      float64
Fri        float64
Student     object
dtype: object


In [204]:
scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8 entries, week_1 to week_4
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Mon      8 non-null      float64
 1   Tue      8 non-null      float64
 2   Wed      8 non-null      float64
 3   Thurs    8 non-null      float64
 4   Fri      8 non-null      float64
 5   Student  8 non-null      object 
dtypes: float64(5), object(1)
memory usage: 448.0+ bytes


#### 2. [P] Show two different ways to report the number of observations in ```scores```.

In [205]:
# method 1: retrieve DataFrame property called size
print("report size via retrieving property .size: " + \
      str(scores.size))

# method 2: multiply 
scores_size = -1
for i in range(len(scores.shape)):
    if i == 0: scores_size = scores.shape[0]
    else: scores_size *= scores.shape[i]
print("report size via multiplying dimensions: " + \
      str(scores_size))
    

report size via retrieving property .size: 48
report size via multiplying dimensions: 48


#### 3. [P] Show the variables (i.e. columns) in ```scores```

In [206]:
# get column index in .column property in DataFrame scores
# to show variables (columns) of scores
scores.columns

Index(['Mon', 'Tue', 'Wed', 'Thurs', 'Fri', 'Student'], dtype='object')

#### 4. [P] There are repetitive names in the index. This happens quite a bit, depending on how the rows are to be named. For now. print the unique index names in ```scores```

In [207]:
scores.index.unique() # first, get the Index object that stores the
                      # the index of the scores DataFrame
                      # then, the call can be made from the
                      # Index object to the .unique function to
                      # return an Index object with the unique
                      # index values from the Index object that
                      # made the call

Index(['week_1', 'week_2', 'week_3', 'week_4'], dtype='object')

#### 5. [P] Rename the 'Thurs' column header to be 'Thu' . (Hint: Changing ```scores.columns[3]``` directly will not work, not to mention it would be a poor approach to hard code a location this way! Magic numbers are never a good idea when referencing index or column locations. They may change!) Show the new scores data frame. This should be the same, except with Thurs changed to Thu.

In [208]:
scores.rename({"Thurs": "Thu"}, axis='columns', inplace=True)
scores

Mon   Tue    Wed    Thu   Fri Student
week_1  9.50  8.75   8.00  10.00  7.75     Bob
week_1  9.00  8.00  10.00   8.75  7.25    Jane
week_2  8.50  0.00   9.75   9.00  6.00     Bob
week_2  7.00  8.25   9.25   0.00  8.00    Jane
week_3  7.00  8.50   9.25   0.00  0.00     Bob
week_3  8.50  8.25   9.00   8.00  7.50    Jane
week_4  9.00  7.00   8.50   8.00  8.00     Bob
week_4  9.25  9.25   8.50   7.75  7.50    Jane

#### 6. [P] Now, make a bigger change. Let's rename '```week_1```' to be '```w1```', '```week_2```' to be 'w2', and so on. Also, suppose you decide to use only 2 letter abbreviations for the days. Rename the days in the column names to be 'Mo', 'Tu', 'We', 'Th', 'Fr'. Show the updated scores dataframe. It should appear as follows:

In [209]:
def _rename_indices_to_w_num(df: pd.DataFrame):
    '''
    Assumption: each row index value in the inputted DataFrame 
    starts with w and ends with the week number
    
    This private utility method renames each row index value in the
    inputted pd.DataFrame object to w<week_number>, e.g. w1 for
    week 1, w2 for week 2, etc.
    
        Parameters:
            df (pd.DataFrame): pd.DataFrame object whose rows
                               are undergoing in the renaming
    '''
    rename_dict = {}
    for index_value in scores.index:
        if not(index_value in rename_dict):
            rename_dict[index_value] = "w" + index_value[-1] 
    df.rename(rename_dict, axis='index', inplace=True)

def _rename_columns_to_two_letter_day_names(df: pd.DataFrame):
    """
    This private utility renames each column name value 
    in the inputted pd.DataFrame object to its two letter 
    equivalent, e.g. Monday or Mon becomes Mo, Tueesday, Tues,
    or Tue becomes Tu, etc.
    
    The implementation sifts through columns in the inputted
    pd.DataFrame object for days of the week. When and if it finds
    a day of the week, it renames it to the two-letter abbreviation
    of that day of the week.
    
        Parameters:
            df (pd.DataFrame): pd.DataFrame object whose columns
                               are undergoing renaming
    """
    
    # list of the variations of names/abbreviations (except the
    # two-letter ones) of the days of the week
    days=['Sunday', 'Monday', 'Tuesday', 'Wednesday', \
          'Thursday', 'Friday', 'Saturday']
    day_abbreviations_without_two_letter=['Sun', 'Mon', 'Tue', \
                                          'Tues', 'Wed', 'Thu', \
                                           'Thurs', 'Fri', 'Sat']
    rename_dict = {}
    for column_value in scores.columns:
        if not(column_value in rename_dict):
            if ((column_value in days) or \
                (column_value in \
                 day_abbreviations_without_two_letter)):
                rename_dict[column_value] = column_value[0:2]
    df.rename(rename_dict, axis='columns', inplace=True)
            
_rename_indices_to_w_num(scores)
_rename_columns_to_two_letter_day_names(scores)

scores

Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50  0.00   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25   0.00  8.00    Jane
w3  7.00  8.50   9.25   0.00  0.00     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane

#### 7. [PM] Compare the type of the expression scores['Mo'] vs. scores[['Mo']]. What is the difference? (HINT: use the type() Python function on both.)

In [210]:
### Borrowed from my responses to lab03

def _print_item_and_type(item_to_print):
    '''
    This local, private utility method prints the item and type of 
    item being printed.
    
        Parameters:
                item_to_print: any time of item being printed
    '''
    try: 
        print(item_to_print)
        print()
        print("return type: " + str(type(item_to_print)))
    except: 
        print("print_type method error: an error occurred \
        while trying to print the type of an object.")
    
def _print_n_lines(num_lines: int):
    '''
    This local, private utility method prints num_lines number 
    of lines.
    
        Parameters:
            num_lines (int): number of lines to be printed
    '''
    for i in range(num_lines):
        print()
        
def _print_two_lines():
    '''
    This local, private utility method calls _print_n_lines to 
    print two lines.
    '''
    _print_n_lines(2)

In [211]:
print("scores['Mo']: \n")
_print_item_and_type(scores['Mo'])
_print_two_lines()
print("scores[['Mo']]: \n")
_print_item_and_type(scores[['Mo']])

scores['Mo']: 

w1    9.50
w1    9.00
w2    8.50
w2    7.00
w3    7.00
w3    8.50
w4    9.00
w4    9.25
Name: Mo, dtype: float64

return type: <class 'pandas.core.series.Series'>


scores[['Mo']]: 

      Mo
w1  9.50
w1  9.00
w2  8.50
w2  7.00
w3  7.00
w3  8.50
w4  9.00
w4  9.25

return type: <class 'pandas.core.frame.DataFrame'>


```scores['Mo']``` extracts the column with the column label 'Mo' from the DataFrame in the form of a *pd.Series object*. On the other hand, ```scores[['Mo']]``` specifies a selection of columns within the DataFrame, which is another *pd.DataFrame*; the selection just happens to only have the column label 'Mo,' hence why the string representations, or print outputs, of ```scores['Mo']``` and ```scores[['Mo']]``` are similar.

#### 8. [P] Demonstrate the describe() method on scores. What type does it return?

In [212]:
_print_item_and_type(scores.describe())

             Mo      Tu         We         Th        Fr
count  8.000000  8.0000   8.000000   8.000000  8.000000
mean   8.468750  7.2500   9.031250   6.437500  6.500000
std    0.967669  3.0000   0.673974   4.037215  2.702512
min    7.000000  0.0000   8.000000   0.000000  0.000000
25%    8.125000  7.7500   8.500000   5.812500  6.937500
50%    8.750000  8.2500   9.125000   8.000000  7.500000
75%    9.062500  8.5625   9.375000   8.812500  7.812500
max    9.500000  9.2500  10.000000  10.000000  8.000000

return type: <class 'pandas.core.frame.DataFrame'>


describe() on scores returns an object that is of type DataFrame (returns a DataFrame). 

It returns the following (*extra information*):
* count: number of values in the column
* mean: mean of values in the column
* std: standard deviation of the values in the column
* min: minimum value of the values in the column
* 25%: 25 percentile of the values in the column
  (numpy.percentile is used to calculate the 
  percentile value<sup>1 2</sup>)
* 50%/median: median of the values in the column
* 75%: 75 percentile of the values in the column
  (numpy.percentile is used to calculate the 
  percentile value<sup>1 2</sup>)
* max: maximum values of the values in the column

Source:
1. https://stackoverflow.com/questions/39567712/python-pandas-how-is-25-percentile-calculated-by-describe-function
2. https://github.com/pandas-dev/pandas/blob/master/pandas/core/series.py

#### 9. [P] Store the output of describe() as a variable. Then, using this data frame, report the day that had the largest standard deviation in quiz scores and its value. Clean your temporary variables when complete (i.e. use del)

#### Your output should look something like:
```Th had the highest standard deviation: 4.037```

In [213]:
class InvalidStatLabelError(Exception):
    """
    Raised when the inputted label is not one of the statistics
    (row indices) outputted from the pd.DataFrame function
    .describe()
    """
    pass

def _print_max_value_and_column_label(df: pd.DataFrame, 
                                     stat_label: str,
                                     stat_full_name: str,
                                     num_dec_plcs: int):
    """
    This private, local utility method prints the maximum 
    value of the attribute stat_full_name abbreviated by stat_label
    (within the index of the pd.DataFrame object) from the inputted
    pd.DataFrame object df. The result is printed to three decimal 
    places by default.
    
        Parameters: 
                df (pd.DataFrame): pd.DataFrame object whose
                                   attributes are being analyzed
                stat_label (str): abbreviation of statistics
                                  within the index of the
                                  inputted pd.DataFrame object
                                  df
                stat_full_name (str): full name of the statistics
                                      whose maximum being searched,
                                      retrieved, and printed
                num_dec_places (int): number of decimals the
                                      printed statistics value
                                      will have
                                      
    """
    df_stats_summary = df.describe() # stats from .describe
                                     # on the inputted pd.DataFrame
                                     # object
    try:
        # isolating the row of stats being selected based on
        # the inputted stat_label, e.g. std, mean, etc.
        selected_stat_row = df_stats_summary.loc[stat_label]
    except KeyError:
        raise InvalidStatLabelError(
            stat_label + " is not a valid " + 
            "row in the inputted DataFrame.")
    except:
        print("Some other exception occurred while trying to " + 
              "access the requested stat in the inputted " + 
              "pd.DataFrame object.\n")
        raise
    
    column_label_of_max_stat = selected_stat_row.idxmax()
    print(
        column_label_of_max_stat + 
        " had the highest " +
        stat_full_name +
        (": {:." + str(num_dec_plcs) + "f}") 
        .format( 
                selected_stat_row.at[column_label_of_max_stat] 
                )
        )
    # deleting temporary function parameter variables
    del df
    del stat_label
    del stat_full_name
    del num_dec_plcs
    # deleting temporary local variables
    del df_stats_summary
    del selected_stat_row
    del column_label_of_max_stat

_print_max_value_and_column_label(scores, 
                                  "std", 
                                  "standard deviation", 
                                  3)

Th had the highest standard deviation: 4.037


#### 10. [P] Knowing that describe() is nothing more than another data frame is interesting. However, it's more important to write code to output the specific results you are looking for. Write the Python code to repeat the previous exercise without using describe().

In [214]:
day_std_values = scores.std(axis=0)
day_max_std = day_std_values.idxmax()
num_dec_plcs = 3

print(
      day_max_std + \
      " had the highest standard deviation: " + \
      ("{:." + str(num_dec_plcs) + "f}") \
       .format(day_std_values.at[day_max_std])
     )


Th had the highest standard deviation: 4.037


#### 11. [P] Write the code that changes all 0.0 entries to np.nan. (NOTE: You should be able to do this with just one line of code using pandas selection techniques!) Then, show scores. All 0.0 entries should be replaced with NaN in the output.

In [215]:
scores.replace(0, np.nan, inplace=True)

scores

Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50   NaN   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25    NaN  8.00    Jane
w3  7.00  8.50   9.25    NaN   NaN     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane

#### 12. [P] Show the output of describe() again. Yes, this is better! Note how much more informative this output is when properly dealing with missing data. Copy your code from above to output the day with the largest standard deviation in quiz scores. You should have a substantially different result than you did before!

In [216]:
print(scores.describe())
print()

day_std_values = scores.std(axis=0)
day_max_std = day_std_values.idxmax()
stat_sig_figs = 3

print(
      day_max_std + \
      " had the highest standard deviation: " + \
      ("{:." + str(stat_sig_figs) + "f}") \
       .format(day_std_values.at[day_max_std])
     )



             Mo        Tu         We         Th        Fr
count  8.000000  7.000000   8.000000   6.000000  7.000000
mean   8.468750  8.285714   9.031250   8.583333  7.428571
std    0.967669  0.698638   0.673974   0.846562  0.687906
min    7.000000  7.000000   8.000000   7.750000  6.000000
25%    8.125000  8.125000   8.500000   8.000000  7.375000
50%    8.750000  8.250000   9.125000   8.375000  7.500000
75%    9.062500  8.625000   9.375000   8.937500  7.875000
max    9.500000  9.250000  10.000000  10.000000  8.000000

Mo had the highest standard deviation: 0.968


#### 13. [PM] Show the output of scores.values. What does the .values attribute do?

In [217]:
print(scores.values)
print()
print("type of scores.values: " + str(type(scores.values)))

[[9.5 8.75 8.0 10.0 7.75 'Bob']
 [9.0 8.0 10.0 8.75 7.25 'Jane']
 [8.5 nan 9.75 9.0 6.0 'Bob']
 [7.0 8.25 9.25 nan 8.0 'Jane']
 [7.0 8.5 9.25 nan nan 'Bob']
 [8.5 8.25 9.0 8.0 7.5 'Jane']
 [9.0 7.0 8.5 8.0 8.0 'Bob']
 [9.25 9.25 8.5 7.75 7.5 'Jane']]

type of scores.values: <class 'numpy.ndarray'>


The .values attribute of a DataFrame object returns a 2D
ndarray object with the underlying values stored in the
DataFrame object, removing the labels of the axes.<sup>1</sup>

Source:
1. https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.values.html

#### 14. [P] What is the mean quiz score for each day? Do not use describe()

In [218]:
scores.mean(axis=0, skipna=True) # skipna=True redundant
                                 # setting intention for
                                 # maintainability

print(scores)

      Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50   NaN   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25    NaN  8.00    Jane
w3  7.00  8.50   9.25    NaN   NaN     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane


#### 15. [P] What is the mean quiz score over the entire dataset, ignoring all missing values? (HINT: there are many ways to do this, and no solution should involve writing any loops! The only right answer is 8.375. If you aren't getting that, then you are not handling NaN values correctly. You only want the mean over all scores that are NOT NaN!)

In [219]:
numeric_values_in_scores = scores.select_dtypes(np.number)
np.nanmean(numeric_values_in_scores.values)

8.375

#### 16. [P] Show the mean for each week over all students. Again, you are not counting missing quizzes into the average. (Yes, technically you could write a loop to compute this exhaustively. However, you should rarely need to do that. Use pandas!)

#### You have the correct result if you report the following:
#### *(same as output below)*

In [294]:
# IMPLEMENTATION #1 (my improvement on IMPLEMENTATION #2)

student_col_name = 'Student' 

print("Mean score for each week over all students")

# specification of which student the scores belong to
# are not relevant for the purposes of this conversation and
# will only interfere with calculating the mean down the road
scores_without_student_names = scores.drop(columns= \
                                           student_col_name)

print(
    scores.index.unique().to_series().apply( # vectorize on each
                                             # index value
        lambda x: scores_without_student_names \
                  .loc[x] # obtain DataFrame selecting week \
                  .stack() # make that DataFrame into a Series \
                  .mean(skipna=True) # find mean of the Series
                                     # while skipping all NaN vals
    )
    
)

Mean score for each week over all students
w1    8.70000
w2    8.21875
w3    8.25000
w4    8.27500
dtype: float64


In [295]:
# IMPLEMENTATION #2 (my first reasonable implementation)

student_col_name = 'Student' 

print("Mean score for each week over all students")
# after removing column with student names, create a group
# of DataFrames by the index, or week, that is
# (represented by a pandas.core.groupby.generic.DataFrameGroupBy
# object)
# each DataFrame (containing a particular week's data) is a 'group'
week_groups = scores. \
                  drop(columns=student_col_name). \
                  groupby(level=0, 
                          dropna=False) 

# finding unique index values, i.e. name of each week like
# w1 for week 1, w2 for week 2, etc. and then converting it to
# a Series object instead of list to later use the vectorization
# methods in pandas
index_values = scores.index.unique().to_series()

# for each unique index value, retrieve a DataFrame containing
# the data associated with that unique index value (one or more
# rows in the original data frame - in this case, scores) using
# function vectorization of pandas.core.groupby.Groupby.get_group
# the result is a series of DataFrames ('groups')
series_of_dfs = scores.index.unique(). \
to_series().apply(week_groups.get_group)

# finally, for each DataFrame in the Series series_of_dfs,
# stack it and then take the mean of that stack, skipping the over
# the NaN values
print(index_values.apply(
    lambda x: lst_of_groups[x].stack().mean(skipna=True))) 
                                            # default for skipna
                                            # is True, but better
                                            # to specify for
                                            # readability

Mean score for each week over all students
w1    8.70000
w2    8.21875
w3    8.25000
w4    8.27500
dtype: float64


#### 17. [P] Show the mean score for each student. Again, ignore all missing values.

In [296]:
student_col_name = 'Student' 

print("Mean score for each student over all the weeks")

# extract student names and then put in Series with
# same indexing as values
student_names_np_arr = scores[student_col_name].unique()
student_names = pd.Series(student_names_np_arr, 
                          index=student_names_np_arr)

print(
    student_names.apply( # vectorize on each student
        
              # pick scores of student
    lambda x: scores.loc[scores[student_col_name]==x,:] \
        .drop(columns=student_col_name) # student column will \
                                       # interfere in calculating \
                                       # the mean \
        .stack() # make that DataFrame into a Series \
        .mean(skipna=True) # find mean of the Series
                           # while skipping all NaN vals
    )
)

Mean score for each student over all the weeks
Bob     8.382353
Jane    8.368421
dtype: float64


#### 18. [P] Select the data frame for only Monday and Student variables using .loc, and again using .iloc

In [303]:
col_names_lst = ["Mo", "Student"] # student names in list
                                  # list for easy modification

col_names = pd.Series(col_names_lst) # student names 
                                     # in pd.Series

print("using .loc: ")
print(scores.loc[:, col_names])
print()
print("using .iloc: ")
print(
    scores.iloc[:, col_names.apply(lambda x: 
                                   scores.columns.get_loc(x))]
)

using .loc: 
      Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane

using .iloc: 
      Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane
